source/common/upstream/load_balancer_impl.cc

```diff
@@ -1299,6 +1299,33 @@ HostConstSharedPtr LeastRequestLoadBalancer::unweightedHostPick(const HostVector
                                                                 const HostsSource&) {
   HostSharedPtr candidate_host = nullptr;
 
+  // Use a full scan if:
+  // - it's explicitly configured, or
+  // - the number of choices is equal to or larger than the number of hosts.
+  const auto has_full_scan_configured = selection_method_ ==
+                                        envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest_SelectionMethod_FULL_SCAN;
+  if (has_full_scan_configured || (hosts_to_use.size() <= choice_count_)) {
+    // Choose a random index to start from preventing always picking the first host in the list.
+    const int rand_idx = random_.random() % hosts_to_use.size();
+    for (unsigned long i = 0; i < hosts_to_use.size(); i++) {
+      const HostSharedPtr& sampled_host = hosts_to_use[(rand_idx + i) % hosts_to_use.size()];
```

Comment:
We don't want to change the default behavior of the LB if FULL_SCAN is not set, so I'd remove the part that checks `hosts_to_use.size() <= choice_count_`. Otherwise, this will break things for a lot of folks that depend on the subtle differences in selection probabilities when using the n-choice selection. 

I think I go into the details in https://github.com/envoyproxy/envoy/pull/31146#pullrequestreview-1763509885.